### DISCUSSION TITLE
Why PullGestureRecognizer doesn't work at all?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#17493] [Q&A]
Here is my code:
```
<Border Width="500"
    Height="500"
    Margin="5"
		Background="Red"
    Name="border">		
</Border>
```

And:
```
 public MainView2()
 {
     InitializeComponent();        
     border.GestureRecognizers.Add(new PullGestureRecognizer()
     {
         PullDirection = PullDirection.TopToBottom,
     });
     border.AddHandler(Gestures.PullGestureEvent, (s, e) =>
     {
     });
     border.AddHandler(Gestures.PullGestureEndedEvent, (s, e) =>
     {
     });
 }
```

And I add breakpoint to PullGestureEvent and PullGestureEndedEvent above.  However, no matter I drag from top to bottom, the breakpoint never works. What's wrong with my code?
![1](https://github.com/user-attachments/assets/bf90b4c0-ab07-40c9-8a16-f2934cb0a122)

**Comment by thevortexcloud** [comment]
I could be wrong, but I seem to vaguely recall that some gesture recognisers are touch only. Which means you need a real touch screen to use them. But I also don't have much experience using them.

**Reply by IanRawley** [reply]
PullGestureRecognizer is Touch or Pen only.  I'm not really sure why?  Maybe to avoid conflicts with drag-and-drop?
ScrollGestureRecognizer as well.

Edit: PullGestureRecognizer was added with RefreshContainer, so possibly just a direct port from somewhere else.

**Reply by kyuranger** [reply]
> I could be wrong, but I seem to vaguely recall that some gesture recognisers are touch only. Which means you need a real touch screen to use them. But I also don't have much experience using them.

I am working on a cross-platform program and I hope that the display is the same on different platforms. In the past, I used Thumb to achieve the drag effect, but using Thumb requires you to add additional controls, while PullGestureRecognizer does not. But as you said, PullGestureRecognizer may not support mouse control.

**Reply by kyuranger** [reply]
> PullGestureRecognizer is Touch or Pen only. I'm not really sure why? Maybe to avoid conflicts with drag-and-drop? ScrollGestureRecognizer as well.
> 
> Edit: PullGestureRecognizer was added with RefreshContainer, so possibly just a direct port from somewhere else.

If this is the case, I can only use Thumb to implement it as I did before.

**Reply by thevortexcloud** [reply]
> I'm not really sure why?

I wish I knew. It's a fairly weird restriction to me and it probably makes testing a lot harder.

**Comment by LaurentInSeattle** [comment]
Maybe... Removing this check in PullGestureRecognizer.cs would make it work with the mouse ? 

protected override void PointerPressed(PointerPressedEventArgs e)
        {
            if (Target != null && Target is Visual visual && (e.Pointer.Type == PointerType.Touch || e.Pointer.Type == PointerType.Pen))
            {

**Reply by kyuranger** [reply]
I tried to Override this method as you said but found that some methods in it are private (such as PullGestureEventArgs.GetNextFreeId()), so it is not feasible at all.

**Comment by LaurentInSeattle** [comment]
The gesture is indeed recognized and the events are fired if: 

-  Duplicate original, edit 
-  you do remove the pointer type test 
-  you add something like :      
    private static int _nextId = 1;
    internal static int GetNextFreeId() => _nextId++;
to replace the internal GetNextFreeId from the event args

Works on my machine :) 

Ugly work around for sure but proves that the Avalonia code could easily be fixed...
